File: src/db/knowledge.rs
//! Knowledge repository for persistent agent facts
//!
//! Allows agents to store and retrieve discovered information (e.g. ports, paths).

use anyhow::{Context, Result};
use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};
use std::sync::{Arc, Mutex};

/// A single unit of knowledge
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Knowledge {
    pub key: String,
    pub value: String,
    pub category: String,
    pub description: Option<String>,
}

/// Repository for knowledge operations
pub struct KnowledgeRepository {
    conn: Arc<Mutex<Connection>>,
}

impl KnowledgeRepository {
    pub fn new(conn: Arc<Mutex<Connection>>) -> Self {
        Self { conn }
    }

    /// Set a fact (insert or replace)
    pub fn set(
        &self,
        key: &str,
        value: &str,
        category: &str,
        description: Option<&str>,
    ) -> Result<()> {
        let conn = self.conn.lock().unwrap();
        conn.execute(
            "INSERT OR REPLACE INTO knowledge (key, value, category, description, updated_at) 
             VALUES (?1, ?2, ?3, ?4, datetime('now'))",
            params![key, value, category, description],
        )
        .context("Failed to set knowledge")?;
        Ok(())
    }

    /// Get a fact by key
    pub fn get(&self, key: &str) -> Result<Option<Knowledge>> {
        let conn = self.conn.lock().unwrap();
        let mut stmt =
            conn.prepare("SELECT key, value, category, description FROM knowledge WHERE key = ?1")?;

        let mut rows = stmt.query(params![key])?;
        if let Some(row) = rows.next()? {
            Ok(Some(Knowledge {
                key: row.get(0)?,
                value: row.get(1)?,
                category: row.get(2)?,
                description: row.get(3)?,
            }))
        } else {
            Ok(None)
        }
    }

    /// List all knowledge, optionally filtered by category
    pub fn list(&self, category_filter: Option<&str>) -> Result<Vec<Knowledge>> {
        let conn = self.conn.lock().unwrap();
        let mut sql = "SELECT key, value, category, description FROM knowledge".to_string();

        if category_filter.is_some() {
            sql.push_str(" WHERE category = ?1");
        }
        sql.push_str(" ORDER BY category, key");

        let mut stmt = conn.prepare(&sql)?;

        // Use a helper closure to map rows
        let mapper = |row: &rusqlite::Row| -> rusqlite::Result<Knowledge> {
            Ok(Knowledge {
                key: row.get(0)?,
                value: row.get(1)?,
                category: row.get(2)?,
                description: row.get(3)?,
            })
        };

        // Execute query
        let iter = if let Some(cat) = category_filter {
            stmt.query_map(params![cat], mapper)?
        } else {
            stmt.query_map([], mapper)?
        };

        let mut result = Vec::new();
        for item in iter {
            result.push(item?);
        }
        Ok(result)
    }

    /// Delete a fact
    pub fn delete(&self, key: &str) -> Result<()> {
        let conn = self.conn.lock().unwrap();
        conn.execute("DELETE FROM knowledge WHERE key = ?1", params![key])?;
        Ok(())
    }

    /// Track a server process by port and PID
    /// Convenience wrapper for common use case of tracking started servers
    pub fn track_server(&self, port: u16, pid: u32) -> Result<()> {
        let key = format!("server_port_{}_pid", port);
        let description = format!("Dev server on port {}", port);
        self.set(&key, &pid.to_string(), "servers", Some(&description))
    }

    /// Get the tracked PID for a server on a given port
    pub fn get_tracked_server(&self, port: u16) -> Result<Option<u32>> {
        let key = format!("server_port_{}_pid", port);
        if let Some(knowledge) = self.get(&key)? {
            Ok(knowledge.value.parse().ok())
        } else {
            Ok(None)
        }
    }

    /// Remove tracking for a server (call after killing it)
    pub fn untrack_server(&self, port: u16) -> Result<()> {
        let key = format!("server_port_{}_pid", port);
        self.delete(&key)
    }
}

File: forger/internal/db/knowledge.go
package db

import (
	"database/sql"
	"time"
)

// Knowledge represents a knowledge base entry
type Knowledge struct {
	Key         string
	Value       string
	Category    string
	Description sql.NullString
	CreatedAt   time.Time
	UpdatedAt   time.Time
}

// KnowledgeRepository handles knowledge base operations
type KnowledgeRepository struct {
	db *sql.DB
}

// NewKnowledgeRepository creates a new knowledge repository
func NewKnowledgeRepository(db *sql.DB) *KnowledgeRepository {
	return &KnowledgeRepository{db: db}
}

// Create creates a new knowledge entry
func (r *KnowledgeRepository) Create(knowledge *Knowledge) error {
	_, err := r.db.Exec(`
		INSERT INTO knowledge (key, value, category, description)
		VALUES (?, ?, ?, ?)
	`, knowledge.Key, knowledge.Value, knowledge.Category, knowledge.Description)
	return err
}

// GetByKey retrieves a knowledge entry by key
func (r *KnowledgeRepository) GetByKey(key string) (*Knowledge, error) {
	var k Knowledge
	var description sql.NullString

	err := r.db.QueryRow(`
		SELECT key, value, category, description, created_at, updated_at
		FROM knowledge WHERE key = ?
	`, key).Scan(
		&k.Key, &k.Value, &k.Category, &description,
		&k.CreatedAt, &k.UpdatedAt,
	)
	if err != nil {
		return nil, err
	}

	k.Description = description
	return &k, nil
}

// GetAll retrieves all knowledge entries
func (r *KnowledgeRepository) GetAll() ([]Knowledge, error) {
	rows, err := r.db.Query(`
		SELECT key, value, category, description, created_at, updated_at
		FROM knowledge ORDER BY category, key
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var knowledge []Knowledge
	for rows.Next() {
		var k Knowledge
		var description sql.NullString

		if err := rows.Scan(
			&k.Key, &k.Value, &k.Category, &description,
			&k.CreatedAt, &k.UpdatedAt,
		); err != nil {
			return nil, err
		}

		k.Description = description
		knowledge = append(knowledge, k)
	}

	if err := rows.Err(); err != nil {
		return nil, err
	}

	return knowledge, nil
}

// Update updates an existing knowledge entry
func (r *KnowledgeRepository) Update(knowledge *Knowledge) error {
	_, err := r.db.Exec(`
		UPDATE knowledge
		SET value = ?, category = ?, description = ?
		WHERE key = ?
	`, knowledge.Value, knowledge.Category, knowledge.Description, knowledge.Key)
	return err
}

// Delete deletes a knowledge entry by key
func (r *KnowledgeRepository) Delete(key string) error {
	_, err := r.db.Exec(`DELETE FROM knowledge WHERE key = ?`, key)
	return err
}
